packages/live-share/src/EphemeralTimer.ts

```diff
@@ -1,241 +1,275 @@
+      const event = this._pauseEvent!.sendEvent({
+        duration: this._currentState.duration,
+        position: this._currentState.position + (EphemeralEvent.getTimestamp() - this._currentState.timestamp)
+      });
 
-    private _handlePlayEvent(event: IPlayPauseEvent, local: boolean) {
-        if (
-            this._state!.duration - this._timerInterval.milliseconds >=
-            this._state!.position
-        ) {
-            this._state = {
-                timeStarted: this._state!.timeStarted,
-                position: event.position,
-                duration: this._state!.duration,
-                running: true,
-            };
-            this._emitState(local);
-            this._handleTimerInterval();
-        } else {
-            this._handleReset(local);
-        }
+      // Update local state immediately
+      // - The _stateUpdatedEvent won't be triggered until the state change is actually sent. If
+      //   the client is disconnected this could be several seconds later.
+      this.updateState(this.pauseEventToState(event), true);
     }
+  }
 
-    private _handlePauseEvent(event: IPlayPauseEvent, local: boolean) {
-        this._state = {
-            timeStarted: this._state!.timeStarted,
-            position: event.position,
-            duration: this._state!.duration,
-            running: true,
-        };
-        this._emitState(local);
-        this._cancelTimerIfRunning();
+  private _handlePlay(event: IPlayEvent, local: boolean) {
+    if (!local) {
+      const newState = this.playEventToState(event);
+      this.remoteStateReceived(newState, event.clientId!);
     }
+  }
 
-    private _handleReset(local: boolean) {
-        this._state = {
-            timeStarted: EphemeralEvent.getTimestamp(),
-            position: 0,
-            duration: this._state!.duration,
-            running: this._state!.running,
-        };
-        this._emitState(local);
-        if (this._state.running) {
-            this._handleTimerInterval();
-        }
+  private _handlePause(event: IPauseEvent, local: boolean) {
+    if (!local) {
+      const newState = this.pauseEventToState(event);
+      this.remoteStateReceived(newState, event.clientId!);
     }
+  }
+
+  private remoteStateReceived(state: ITimerState, sender: string): void {
+    EphemeralEvent.verifyRolesAllowed(sender, this._allowedRoles).then((allowed) => {
+      // Ensure that state is allowed, newer, and not the initial state.
+      if (allowed && isNewer(this._currentState, state) && state.timestamp !== 0) {
+          this.updateState(state, false);
+      }
+    }).catch((err) => {
+      console.error(err);
+    });
+  }
 
-    private _emitState(local: boolean) {
-        const newState = Object.assign({}, this._state);
-        this.emit("valueChanged", newState, local);
+  private updateState(state: ITimerState, local: boolean) {
+    const clone = cloneValue(state)!;
+    if (!local) {
+      clone.clientId = this._currentState.clientId;
```

Comment:
>clone.clientId = this._currentState.clientId; [](http://example.com/codeflow?start=6&length=45)

why are you overriding the local clientId first?

Reply:
misunderstanding i had last week that got carried over I think